File: php/wallet.php
<?php

$portefeuille = [
    'solde' => 0,
    'totalDepenses' => 0,
];

$transactions = [];
$nombreRetraits = 0;
$cinList = [];

function ajouterCIN($numeroCIN, $nomTitulaire) {
    global $cinList;
    $cin = [
        'numero' => $numeroCIN,
        'nom' => $nomTitulaire,
    ];
    $cinList[] = $cin;
    echo "La CIN $numeroCIN de $nomTitulaire a été ajoutée.\n";
}

function ajouterArgent($montant) {
    global $portefeuille, $transactions;
    if ($montant <= 0) {
        echo "Le montant doit être supérieur ou égal à 0.\n";
    } else {
        $portefeuille['solde'] += $montant;
        enregistrerTransaction("Dépôt", $montant);
        echo "Le solde du portefeuille est maintenant de {$portefeuille['solde']} ariary.\n";
    }
}

function retirerArgent($montant) {
    global $portefeuille, $transactions, $nombreRetraits;
    if ($montant <= 0) {
        echo "Le montant doit être supérieur ou égal à 0.\n";
    } elseif ($montant > $portefeuille['solde']) {
        echo "Le montant à retirer est supérieur au solde du portefeuille.\n";
    } elseif ($nombreRetraits >= 3) {
        echo "Limite de retraits atteinte. Vous ne pouvez pas effectuer plus de 3 retraits.\n";
    } else {
        $portefeuille['solde'] -= $montant;
        enregistrerTransaction("Retrait", $montant);
        $nombreRetraits++;
        echo "Le solde du portefeuille est maintenant de {$portefeuille['solde']} ariary.\n";
    }
}

function afficherSolde() {
    global $portefeuille;
    echo "Le solde du portefeuille est de {$portefeuille['solde']} ariary.\n";
}

function afficherHistorique() {
    global $transactions;
    echo "\nHistorique des transactions :\n";
    foreach ($transactions as $transaction) {
        echo "{$transaction['type']} de {$transaction['montant']} ariary le {$transaction['date']}\n";
    }
    echo "\n";
}

function enregistrerTransaction($type, $montant) {
    global $transactions, $portefeuille;
    $transaction = [
        'type' => $type,
        'montant' => $montant,
        'date' => date("Y-m-d H:i:s"),
    ];
    $transactions[] = $transaction;

    if ($type === "Retrait") {
        $portefeuille['totalDepenses'] += $montant;
    }
}

function afficherCINList() {
    global $cinList;
    echo "\nListe des CIN enregistrées :\n";
    foreach ($cinList as $cin) {
        echo "CIN {$cin['numero']} - Titulaire: {$cin['nom']}\n";
    }
    echo "\n";
}

function afficherTotalDepenses() {
    global $portefeuille;
    echo "Le total des dépenses est de {$portefeuille['totalDepenses']} ariary.\n";
}

$choix = 0;

while ($choix !== 7) {
    echo "Que voulez-vous faire ?\n";
    echo "1. Ajouter de l'argent\n";
    echo "2. Retirer de l'argent\n";
    echo "3. Afficher le solde\n";
    echo "4. Afficher l'historique des transactions\n";
    echo "5. Ajouter une CIN\n";
    echo "6. Afficher le total des dépenses\n";
    echo "7. Quitter\n";

    $choix = (int) readline("Votre choix : ");

    switch ($choix) {
        case 1:
            $montantAjout = (int) readline("Montant à ajouter : ");
            ajouterArgent($montantAjout);
            break;
        case 2:
            $montantRetrait = (int) readline("Montant à retirer : ");
            retirerArgent($montantRetrait);
            break;
        case 3:
            afficherSolde();
            break;
        case 4:
            afficherHistorique();
            break;
        case 5:
            $numeroCIN = readline("Numéro de CIN : ");
            $nomTitulaire = readline("Nom du titulaire : ");
            ajouterCIN($numeroCIN, $nomTitulaire);
            break;
        case 6:
            afficherTotalDepenses();
            break;
        case 7:
            echo "Au revoir !\n";
            break;
        default:
            echo "Choix incorrect.\n";
    }
}
?>

File: python/wallet.py
import datetime

# Déclaration du portefeuille
portefeuille = {
    'solde': 0,
    'total_depenses': 0,
}

transactions = []
nombre_retraits = 0
cin_list = []

# Ajouter une CIN
def ajouter_cin(numero_cin, nom_titulaire):
    cin = {
        'numero': numero_cin,
        'nom': nom_titulaire,
    }
    cin_list.append(cin)
    print(f"La CIN {numero_cin} de {nom_titulaire} a été ajoutée.")

# Ajouter de l'argent
def ajouter_argent(montant):
    global nombre_retraits
    if montant <= 0:
        print("Le montant doit être supérieur ou égal à 0.")
    else:
        # Augmenter le solde
        portefeuille['solde'] += montant
        enregistrer_transaction("Dépôt", montant)
        print(f"Le solde du portefeuille est maintenant de {portefeuille['solde']} ariary.")

# Retirer de l'argent
def retirer_argent(montant):
    global nombre_retraits
    if montant <= 0:
        print("Le montant doit être supérieur ou égal à 0.")
    elif montant > portefeuille['solde']:
        print("Le montant à retirer est supérieur au solde du portefeuille.")
    elif nombre_retraits >= 3:
        print("Limite de retraits atteinte. Vous ne pouvez pas effectuer plus de 3 retraits.")
    else:
        # enregistrer la transaction et mettre à jour le nombre de retraits
        portefeuille['solde'] -= montant
        enregistrer_transaction("Retrait", montant)
        nombre_retraits += 1
        print(f"Le solde du portefeuille est maintenant de {portefeuille['solde']} ariary.")

# Afficher le solde
def afficher_solde():
    print(f"Le solde du portefeuille est de {portefeuille['solde']} ariary.")

# Afficher l'historique
def afficher_historique():
    print("\nHistorique des transactions :")
    for transaction in transactions:
        print(f"{transaction['type']} de {transaction['montant']} ariary le {transaction['date']}")
    print("\n")

# Enregistrer une transaction dans l'historique
def enregistrer_transaction(type_transaction, montant):
    global portefeuille
    transaction = {
        'type': type_transaction,
        'montant': montant,
        'date': datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
    }
    transactions.append(transaction)

    # Mettre à jour le total des dépenses en fonction du type de transaction
    if type_transaction == "Retrait":
        portefeuille['total_depenses'] += montant

# Afficher la liste des CIN
def afficher_cin_list():
    print("\nListe des CIN enregistrées :")
    for cin in cin_list:
        print(f"CIN {cin['numero']} - Titulaire: {cin['nom']}")
    print("\n")

# Afficher le total des dépenses
def afficher_total_depenses():
    print(f"Le total des dépenses est de {portefeuille['total_depenses']} ariary.")

choix = 0

while choix != 7:
    print("Que voulez-vous faire ?")
    print("1. Ajouter de l'argent")
    print("2. Retirer de l'argent")
    print("3. Afficher le solde")
    print("4. Afficher l'historique des transactions")
    print("5. Ajouter une CIN")
    print("6. Afficher le total des dépenses")
    print("7. Quitter")

    choix = int(input("Votre choix : "))

    if choix == 1:
        montant_ajout = int(input("Montant à ajouter : "))
        ajouter_argent(montant_ajout)
    elif choix == 2:
        montant_retrait = int(input("Montant à retirer : "))
        retirer_argent(montant_retrait)
    elif choix == 3:
        afficher_solde()
    elif choix == 4:
        afficher_historique()
    elif choix == 5:
        numero_cin = input("Numéro de CIN : ")
        nom_titulaire = input("Nom du titulaire : ")
        ajouter_cin(numero_cin, nom_titulaire)
    elif choix == 7:
        print("Au revoir !")
    elif choix == 6:
        afficher_total_depenses()
    else:
        print("Choix incorrect.")